MEN-7169 - feat: added multi sorting capabilities to devices view

mzedel opened the pull request.

alfrunes reviewed (commented):
Just sharing some thoughts, please take with a grain of :salt:
My hope is to deal with some of the technical debt we have in our QA infrastructure, in this case how the test environment so tied to the numerous compose files and overrides we have in the intergration repository as well as the exposed env variables in the backend tests environment makes it nearly impossible to reproduce the CI environment on the local machine.
To give an example, the `testutils/container_manager.DockerComposeManager` uses a mix of `docker` (CLI) and `grep` as well as the Docker python library, and mostly ignoring the compose labels for container lookups. On top of that, there are all the hard-coded compose paths as constants inside the library.

However, zooming out a bit, I don't think it's possible to address all of the issues within the scope of the 13 SP for this task. I would rather prioritize importing the tests, and making them run in the reference environment in `/dev/docker-compose.yml` and if possible remove the scripted entrypoint and replace it with a simple `pytest <...> path/to/tests`.

tranchitella commented on backend/integration-tests/README.md:
````diff
@@ -0,0 +1,16 @@
+
+```bash
+$ ./run -s enterprise
+```
````
This will be a follow-up, for now let's keep the flag.

tranchitella commented on backend/integration-tests/bin/run:
this will be a follow-up

tranchitella commented on backend/tests/integration/docker/docker-compose.storage.azblob.yml:
```diff
@@ -0,0 +1,15 @@
+      image: alpine
+      command: "true"
+    mender-deployments:
+        # HACK: Changing minio dependency condiditon to service_started allows
```
outdated

tranchitella commented on backend/tests/integration/docker/docker-compose.storage.azblob.yml:
```diff
@@ -0,0 +1,15 @@
+services:
+    minio:
+      image: alpine
+      command: "true"
```
outdated

tranchitella commented on backend/tests/integration/docker/stripe-test.docker-compose.yml:
```diff
@@ -0,0 +1,8 @@
+            TENANTADM_STRIPE_API_KEY: ${STRIPE_API_KEY}
+
+networks:
+    mender:
```
outdated

tranchitella commented on backend/tests/integration/docker/stripe-test.docker-compose.yml:
```diff
@@ -0,0 +1,8 @@
+services:
+
+    mender-tenantadm:
```
outdated

tranchitella commented on backend/tests/integration/docker/tenantadm-test-recaptcha-conf.yml:
```diff
@@ -0,0 +1,8 @@
+            TENANTADM_RECAPTCHA_SECRET: 6LeIxAcTAAAAAGG-vFI1TnRWxMZNFuojJ4WifJWe
+
+networks:
+    mender:
```
outdated

tranchitella commented on backend/tests/integration/docker/workflows-worker-smtp-test.yml:
```diff
@@ -0,0 +1,9 @@
+            - WORKFLOWS_SMTP_HOST=mender-backend-tests-runner:4444
+
+networks:
+    mender:
```
outdated

tranchitella commented on backend/tests/integration/docker/workflows-worker-smtp-test.yml:
```diff
@@ -0,0 +1,9 @@
+services:
+
+    mender-workflows-worker:
```
outdated

tranchitella commented on backend/tests/integration/tests/test_api_endpoints.py:
```diff
@@ -0,0 +1,174 @@
+        "useradm",
+        "workflows",
+    )
+
```
Before, this test was checking out every single repo to get the API docs; with the new implementation, we'll use the API docs already in the repo.

tranchitella commented on backend/tests/integration/tests/testutils/infra/cli.py:
```diff
@@ -0,0 +1,274 @@
+class CliTenantadm(BaseCli):
+    def __init__(self, containers_namespace="backend-tests", container_manager=None):
+        BaseCli.__init__(
+            self, "mender-tenantadm", containers_namespace, container_manager
```
outdated

tranchitella commented on backend/tests/integration/tests/testutils/common.py:
```diff
@@ -0,0 +1,607 @@
+    tenantadm_host = (
+        tenantadm.HOST
+        if isK8S() or container_manager is None
+        else container_manager.get_ip_of_service("mender-tenantadm")[0] + ":8080"
```
outdated

tranchitella commented on backend/tests/integration/docker/docker-compose.storage.azblob.yml:
```diff
@@ -0,0 +1,15 @@
+        # HACK: Changing minio dependency condiditon to service_started allows
+        #       us to scale the service to 0.
+        depends_on:
+          minio:
```
outdated

tranchitella commented on backend/tests/integration/docker/docker-compose.backend-tests.yml:
```diff
@@ -0,0 +1,32 @@
+            - /var/run/docker.sock:/var/run/docker.sock
+        command: ${PYTEST_ARGS}
+        environment:
+            PYTHONPATH: "/tests"
+            PYTHONDONTWRITEBYTECODE: 1
+            TESTS_IN_PARALLEL: ${TESTS_IN_PARALLEL}
+            TENANTADM_STRIPE_API_KEY: ${STRIPE_API_KEY}
+            GATEWAY_HOSTNAME: ${GATEWAY_HOSTNAME}
+            AWS_ACCESS_KEY_ID: ${AWS_ACCESS_KEY_ID}
+            AWS_SECRET_ACCESS_KEY: ${AWS_SECRET_ACCESS_KEY}
+            AWS_DEFAULT_REGION: ${AWS_DEFAULT_REGION}
+            AWS_EKS_CLUSTER_NAME: ${AWS_EKS_CLUSTER_NAME}
+            AWS_IOTCORE_ACCESS_KEY_ID: ${AWS_IOTCORE_ACCESS_KEY_ID}
+            AWS_IOTCORE_SECRET_ACCESS_KEY: ${AWS_IOTCORE_SECRET_ACCESS_KEY}
+            AWS_IOTCORE_REGION: ${AWS_IOTCORE_REGION}
+            AWS_IOTCORE_DEVICE_POLICY_NAME: ${AWS_IOTCORE_DEVICE_POLICY_NAME}
+            AZURE_IOTHUB_CONNECTIONSTRING_B64: ${AZURE_IOTHUB_CONNECTIONSTRING_B64}
+            AZURE_IOTHUB_MOCK: ${AZURE_IOTHUB_MOCK}
+            GMAIL_ADDRESS: ${GMAIL_ADDRESS}
+            GMAIL_PASSWORD: ${GMAIL_PASSWORD}
+            K8S: ${K8S}
```
The needed variables to run all the tests locally are:
- AWS_IOTCORE_DEVICE_POLICY_NAME
- AWS_IOTCORE_ACCESS_KEY_ID
- AWS_IOTCORE_SECRET_ACCESS_KEY
- AWS_IOTCORE_REGION
- AZURE_IOTHUB_CONNECTIONSTRING_B64
- STRIPE_API_KEY

Everything else is required to run the tests in staging (something I'm completely ignoring as we speak).

tranchitella replied:
I made those tests optional (skip if we don't have the env variable set), see the commit.

tranchitella commented on backend/tests/integration/tests/test_account_suspension.py:
```diff
@@ -0,0 +1,234 @@
+        )
+        assert r.status_code == 200
+
+        time.sleep(10)
+
+        # try requesting auth
+        body, sighdr = deviceauth.auth_req(
+            aset.id_data,
+            aset.pubkey,
+            aset.privkey,
+            tenants_users_devices[0].tenant_token,
+        )
+
+        r = dacd.call("POST", deviceauth.URL_AUTH_REQS, body, headers=sighdr)
```
Well, I didn't touch this part...
https://github.com/mendersoftware/integration/blob/master/backend-tests/tests/test_account_suspension.py#L175

alfrunes commented on dev/docker-compose.yml:
```diff
@@ -108,7 +108,7 @@ services:
     command: [server, --automigrate]
     environment:
-      DEVICECONFIG_INVENTORY_URI: http://inventory:8080
+      DEVICECONFIG_INVENTORY_ADDR: http://inventory:8080
```
If I'm not mistaken, this should be `_URI`: https://github.com/mendersoftware/deviceconfig/blob/master/config/config.go#L59
```suggestion
      DEVICECONFIG_INVENTORY_URI: http://inventory:8080
```

alfrunes commented on dev/docker-compose.enterprise.yml:
```diff
@@ -140,7 +142,7 @@ services:
       - mongo
     environment:
-      DEVICECONFIG_INVENTORY_URI: http://inventory:8080
+      DEVICECONFIG_INVENTORY_ADDR: http://inventory:8080
```
See my previous comment.

alfrunes commented on overlay/backend/Makefile:
```diff
@@ -25,7 +25,10 @@ $(TEST_TARGETS):
 	@$(MAKE) -C pkg/
-	
+
+.PHONY: integration-test
+integration-test:
```
Sorry for being picky, but could we change the target to `test-integration` for the sake of consistency. If we created a directory `backend/services/integration` that would create a conflicting target.
```suggestion
.PHONY: test-integration
test-integration:
```

alfrunes commented on backend/tests/integration/tests/test_aws_iotcore.py:
```diff
@@ -0,0 +1,443 @@
+
+
+class TestAWSIoTCoreIntegrationsEnterprise(_TestAWSIoTCoreBase):
+    @pytest.mark.parametrize(
```
This test also depends on the environment variables.
```suggestion
    @pytest.mark.skipif(
        not bool(os.environ.get("AWS_IOTCORE_ACCESS_KEY_ID")),
        reason="AWS_IOTCORE_ACCESS_KEY_ID not provided",
    )
    @pytest.mark.parametrize(
```

alfrunes commented on backend/tests/integration/tests/test_aws_iotcore.py:
```diff
@@ -0,0 +1,443 @@
+
+
+class TestAWSIoTCoreIntegrations(_TestAWSIoTCoreBase):
+    @pytest.mark.parametrize(
```
```suggestion
    @pytest.mark.skipif(
        not bool(os.environ.get("AWS_IOTCORE_ACCESS_KEY_ID")),
        reason="AWS_IOTCORE_ACCESS_KEY_ID not provided",
    )
    @pytest.mark.parametrize(
```

alfrunes commented:
Some of the tests still fail for me

<details>

```
FAILED tests/test_api_endpoints.py::TestAPIEndpointsEnterprise::test_api_endpoints[internal-False-get-https-hosted.mender.io-api/internal/v1/devicemonitor/tenants/{tenant_id}/devices] - requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='traefik', port=443): Read timed out. (read timeout=2.0)
FAILED tests/test_api_endpoints.py::TestAPIEndpointsEnterprise::test_api_endpoints[management-True-get-https-hosted.mender.io-/api/management/v1/useradm/auth/magic/{id}] - requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='traefik', port=443): Read timed out. (read timeout=2.0)
FAILED tests/test_api_endpoints.py::TestAPIEndpointsEnterprise::test_api_endpoints[management-False-get-https-hosted.mender.io-/api/management/v1/useradm/oauth2/{provider}] - requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='traefik', port=443): Read timed out. (read timeout=2.0)
FAILED tests/test_artifact.py::TestUploadArtifactEnterprise::test_upload_artifact_selection_no_match_wrong_depends - assert 200 == 204
FAILED tests/test_artifact.py::TestUploadArtifactEnterprise::test_upload_artifact_selection_already_installed - assert 200 == 204
FAILED tests/test_aws_iotcore.py::TestAWSIoTCoreIntegrationsEnterprise::test_get_and_set[expected_integration0] - assert 400 == 201
FAILED tests/test_devauth.py::TestAuthsetMgmtEnterprise::test_put_status_accept - AssertionError: waiting for state 'noauth' in inventory timed out
FAILED tests/test_devauth.py::TestAuthReqEnterprise::test_submit_accept - assert 504 == 401
FAILED tests/test_inventory.py::TestGetDevicesEnterprise::test_get_devices_ok - assert 29 == (0 + 40)
FAILED tests/test_inventory.py::TestGetDevicesEnterprise::test_filter_devices_ok - assert 35 == (0 + 40)
FAILED tests/test_rbac.py::TestRBACDeploymentsToGroupEnterprise::test_deploy_to_group[test_case1] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACv2DeploymentsEnterprise::test_deploy_to_devices[test_case5] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACv2DeploymentsToGroupEnterprise::test_get_configuration[test_case5] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACv2DeploymentsToGroupEnterprise::test_access_to_devices[test_case0] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACv2DeploymentsToGroupEnterprise::test_access_to_devices[test_case2] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACReleasesEnterprise::test_read_releases_and_artifacts[test_case0] - AssertionError: assert 2 == 0
FAILED tests/test_rbac_v2.py::TestRBACReleasesEnterprise::test_read_releases_and_artifacts[test_case1] - AssertionError: assert 2 == 0
FAILED tests/test_rbac_v2.py::TestRBACReleasesEnterprise::test_manage_releases[test_case0] - AssertionError: assert 2 == 0
======================================================================== 18 failed, 454 passed, 9 skipped, 295 deselected, 25982 warnings in 2376.15s (0:39:36) ========================================================================
```

</details>

Yet, some of them pass (8 out of 18) once I retry, so I guess some of them are just flaky?

<details>

```
FAILED tests/test_api_endpoints.py::TestAPIEndpointsEnterprise::test_api_endpoints[internal-False-get-https-hosted.mender.io-api/internal/v1/devicemonitor/tenants/{tenant_id}/devices] - requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='traefik', port=443): Read timed out. (read timeout=2.0)
FAILED tests/test_api_endpoints.py::TestAPIEndpointsEnterprise::test_api_endpoints[management-True-get-https-hosted.mender.io-/api/management/v1/useradm/auth/magic/{id}] - requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='traefik', port=443): Read timed out. (read timeout=2.0)
FAILED tests/test_aws_iotcore.py::TestAWSIoTCoreIntegrationsEnterprise::test_get_and_set[expected_integration0] - assert 400 == 201
FAILED tests/test_inventory.py::TestGetDevicesEnterprise::test_get_devices_ok - assert 39 == (0 + 40)
FAILED tests/test_inventory.py::TestGetDevicesEnterprise::test_filter_devices_ok - assert 34 == (0 + 40)
FAILED tests/test_rbac.py::TestRBACDeploymentsToGroupEnterprise::test_deploy_to_group[test_case1] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACv2DeploymentsEnterprise::test_deploy_to_devices[test_case5] - assert 404 == 204
FAILED tests/test_rbac_v2.py::TestRBACReleasesEnterprise::test_read_releases_and_artifacts[test_case0] - AssertionError: assert 2 == 0
FAILED tests/test_rbac_v2.py::TestRBACReleasesEnterprise::test_read_releases_and_artifacts[test_case1] - AssertionError: assert 2 == 0
FAILED tests/test_rbac_v2.py::TestRBACReleasesEnterprise::test_manage_releases[test_case0] - AssertionError: assert 2 == 0
================================================================================ 10 failed, 8 passed, 758 deselected, 67 warnings in 198.05s (0:03:18) =================================================================================
```

</details>

tranchitella commented:
> Some of the tests still fail for me

No idea, tests are passing here locally and also in Gitlab (see the other PR where I'm already running the integration tests).
Can we follow-up on the test failures, maybe?

mzedel committed:
chore: added unified sort icon to increase alignment in the UI & support asc/desc/none sorting

Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
fix: fixed regression that broke state based navigation for all post-init interactions

Ticket: None
Changelog: None
Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
feat: added url support for multiple sorting criteria

Ticket: MEN-7169
Changelog: None
Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
chore: aligned sort state shape with updated url handling where possible

Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
chore: aligned search view with reality of absent sorting powers

Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
chore: adopted unified sort icon where easily possible

Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
feat: added multi sorting capabilities to devices view

+ made alternative data table work with multi sorting data to help work with sorting info from url state

Ticket: MEN-7169
Changelog: Title
Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
chore: aligned auditlog sorting indication with rest of the UI

Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

mzedel committed:
chore: cleaned up remnants of artifact_name deprecation

Signed-off-by: Manuel Zedel <manuel.zedel@northern.tech>

aleksandrychev reviewed (approved):
looks good

michaelntech reviewed (changes requested):
I think is still missing some visual indicators to show the user what the current sorting is and how to escape it/start a new sort – just by clicking on columns, I get a bit lost and can't revert back a simple sort on one column.  Will try to provide some UX design so it can be incorporated to this PR
